src/main/java/org/killbill/billing/plugin/dao/payment/PluginPaymentDao.java

```diff
@@ -159,9 +160,57 @@ public RESP_R withConnection(final Connection conn) throws SQLException {
-        final String ccNumber = getProperty(PluginPaymentPluginApi.PROPERTY_CC_NUMBER, properties);
-        final String ccLast4 = ccNumber == null ? null : ccNumber.substring(ccNumber.length() - 5, ccNumber.length() - 1);
 
+        /* Clone our properties, what we have been given might be unmodifiable */
+        final Map<String, String> clonedProperties = new HashMap<>(properties);
+
+        /* Extract known values from the properties map */
+        final String token               = getProperty(PluginPaymentPluginApi.PROPERTY_TOKEN,                 clonedProperties);
+        final String ccFirstName         = getProperty(PluginPaymentPluginApi.PROPERTY_CC_FIRST_NAME,         clonedProperties);
+        final String ccLastName          = getProperty(PluginPaymentPluginApi.PROPERTY_CC_LAST_NAME,          clonedProperties);
+        final String ccType              = getProperty(PluginPaymentPluginApi.PROPERTY_CC_TYPE,               clonedProperties);
+        final String ccExpirationMonth   = getProperty(PluginPaymentPluginApi.PROPERTY_CC_EXPIRATION_MONTH,   clonedProperties);
+        final String ccExpirationYear    = getProperty(PluginPaymentPluginApi.PROPERTY_CC_EXPIRATION_YEAR,    clonedProperties);
+        final String ccNumber            = getProperty(PluginPaymentPluginApi.PROPERTY_CC_NUMBER,             clonedProperties);
+        final String ccStartMonth        = getProperty(PluginPaymentPluginApi.PROPERTY_CC_START_MONTH,        clonedProperties);
+        final String ccStartyear         = getProperty(PluginPaymentPluginApi.PROPERTY_CC_START_YEAR,         clonedProperties);
+        final String ccIssueNumber       = getProperty(PluginPaymentPluginApi.PROPERTY_CC_ISSUE_NUMBER,       clonedProperties);
+        final String ccVerificationValue = getProperty(PluginPaymentPluginApi.PROPERTY_CC_VERIFICATION_VALUE, clonedProperties);
+        final String ccTrackData         = getProperty(PluginPaymentPluginApi.PROPERTY_CC_TRACK_DATA,         clonedProperties);
+        final String address1            = getProperty(PluginPaymentPluginApi.PROPERTY_ADDRESS1,              clonedProperties);
+        final String address2            = getProperty(PluginPaymentPluginApi.PROPERTY_ADDRESS2,              clonedProperties);
+        final String city                = getProperty(PluginPaymentPluginApi.PROPERTY_CITY,                  clonedProperties);
+        final String state               = getProperty(PluginPaymentPluginApi.PROPERTY_STATE,                 clonedProperties);
+        final String zip                 = getProperty(PluginPaymentPluginApi.PROPERTY_ZIP,                   clonedProperties);
+        final String country             = getProperty(PluginPaymentPluginApi.PROPERTY_COUNTRY,               clonedProperties);
+
+        /* Remove the keys from the map that will become the "additional data" */
+        clonedProperties.remove(PluginPaymentPluginApi.PROPERTY_TOKEN);
```

Comment:
Hmm... `Map.remove(key)` already removes the key and returns the value that was associated with it (or `null`), so I'd say it's just a matter of doing something like:

```java
final String token = clonedProperties.remove(PluginPaymentPluginApi.PROPERTY_TOKEN);
```